### DISCUSSION TITLE
What is the magic behind `node:fs`

### DISCUSSION THREAD

**Discussion by xgqfrms** [nodejs/help#5037] [Q&A]
> Is it only supported by built-in modules?

##  demo

> build-in  ✅

```js
// fs-test.cjs
const fs = require('fs');
const node_fs = require('node:fs');

log(`fs =`, fs)
log(`\nnode:fs =`, node_fs)
log(`\nnode_fs.existsSync =`, node_fs.existsSync)

```

```js
// fs-test.mjs
import fs from 'fs';
import node_fs from 'node:fs';
import {existsSync} from 'node:fs';

log(`fs =`, fs)
log(`\nnode:fs =`, node_fs)
log(`\nexistsSync =`, existsSync)

```

> custom ❌

```js
// my-fs.js
function existsSync(path) {
  console.log(`✅ path =`, path ?? `❌`)
}

let my_fs = {};

module.exports = my_fs = {
  // ...
  existsSync,
}

```

But it does not work, what's wrong with that?

```js
// test.js
const my_fs = require('./my-fs')
// build-in module only ❓
const existsSync = require('./my-fs:existsSync')
// ❌ Error: Cannot find module './my-fs:existsSync'

const log = console.log;
log(`\my_fs.existsSync =`, my_fs.existsSync)
// log(`\my_fs:existsSync =`, existsSync)

```

![image](https://user-images.githubusercontent.com/7291672/201471839-2341f354-7fc0-4bb2-ad3d-10424cf091aa.png)

## refs

https://github.com/nodejs/node/blob/v19.0.1/lib/fs.js#L281

**Comment by xgqfrms** [comment]
# namespace ? 

```mjs
// my-fs.mjs
function existsSync(path) {
  console.log(`✅ path =`, path ?? `❌`)
}

const my_fs = {
  existsSync,
};

export default my_fs;
export {
  existsSync,
}

```

```js
// test.mjs
import my_fs from './my-fs.mjs';

// namespace ❓ node
// import node_fs from 'node:fs';
// import node_my_fs from './my-fs.mjs:existsSync';

import {existsSync} from './my-fs.mjs';

const log = console.log;

log(`\my_fs.existsSync =`, my_fs.existsSync)
log(`\my_fs:existsSync =`, existsSync)


// $ node ./test.mjs


```

**Comment by aduh95** [comment]
`node:fs` refers to the built-in module to interact with the local File System. For historical reasons it's also available without the `node:` prefix (i.e. `require('node:fs') === require('fs')`), but don't let that confuse you, it's just an alias, there's no "magic" behind it. 

I think what you are trying to do is:

```js
// test.js
const my_fs = require('./my-fs')
const { existsSync } = require('./my-fs')

const log = console.log;
log(`\my_fs.existsSync =`, my_fs.existsSync)
log(`\my_fs:existsSync =`, existsSync)
```

**Comment by xgqfrms** [comment]
https://nodejs.org/api/esm.html#data-imports

<img width="1411" alt="image" src="https://user-images.githubusercontent.com/7291672/205118691-e341ebb5-fa82-47ea-8285-955bacf65478.png">

## refs

https://github.com/web-full-stack/nextjs-ssr/issues/11

https://github.com/webpack/webpack/issues/13290

**Reply by GabenGar** [reply]
Do you know what `URL` is? `node:` protocol is just a hint for nodejs runtime (and bundlers) that module name string is a built-in module. It will never be available for userland imports.
